File: apps/web/lib/server/lsq.ts
/**
 * Least-squares traverse adjustment (parametric / observation-equation method).
 * Ported from services/engine/app/cogo/lsq.py — numpy replaced with small,
 * self-contained matrix helpers (the systems here are tiny).
 *
 * Unknowns are the coordinates of the free stations. Each leg contributes a
 * distance and a direction observation, giving a closed/link traverse the
 * redundancy that least squares distributes optimally.
 */

import type { Point } from "./geometry";

export interface LsqResidual {
  leg: number;
  from: string | null;
  to: string | null;
  v_distance: number;
  v_direction_sec: number;
  magnitude: number;
}

export interface LsqResult {
  points: Point[];
  residuals: LsqResidual[];
  sigma0: number;
  iterations: number;
}

interface LsqLeg {
  bearing: number; // decimal degrees
  distance: number; // metres
  from_name?: string | null;
  to_name?: string | null;
}

function wrap(rad: number): number {
  while (rad > Math.PI) rad -= 2 * Math.PI;
  while (rad < -Math.PI) rad += 2 * Math.PI;
  return rad;
}

/** Solve a (square) linear system N·x = rhs via Gaussian elimination w/ partial
 *  pivoting. Near-singular pivots leave that unknown at 0 (graceful fallback). */
function solveLinear(N: number[][], rhs: number[]): number[] {
  const n = rhs.length;
  // Augmented matrix copy.
  const a = N.map((row, i) => [...row, rhs[i]]);
  for (let col = 0; col < n; col++) {
    // Partial pivot.
    let pivot = col;
    for (let r = col + 1; r < n; r++) {
      if (Math.abs(a[r][col]) > Math.abs(a[pivot][col])) pivot = r;
    }
    if (Math.abs(a[pivot][col]) < 1e-15) continue; // singular column -> skip
    [a[col], a[pivot]] = [a[pivot], a[col]];
    const div = a[col][col];
    for (let j = col; j <= n; j++) a[col][j] /= div;
    for (let r = 0; r < n; r++) {
      if (r === col) continue;
      const factor = a[r][col];
      if (factor === 0) continue;
      for (let j = col; j <= n; j++) a[r][j] -= factor * a[col][j];
    }
  }
  return a.map((row) => row[n]);
}

export function leastSquaresAdjust(
  start: Point,
  legs: LsqLeg[],
  provisional: Point[],
  closed: boolean,
  endKnown: Point | null,
  sigmaDist = 0.01,
  sigmaDirSec = 5.0
): LsqResult {
  const k = legs.length;
  const coords: [number, number][] = provisional.map((p) => [p.east, p.north]);

  const fixed = new Set<number>([0]);
  if (closed) {
    coords[k] = [start.east, start.north];
    fixed.add(k);
  } else if (endKnown !== null) {
    coords[k] = [endKnown.east, endKnown.north];
    fixed.add(k);
  }

  const free: number[] = [];
  for (let i = 0; i <= k; i++) if (!fixed.has(i)) free.push(i);
  const col = new Map<number, number>();
  free.forEach((st, j) => col.set(st, j));
  const u = 2 * free.length;

  if (u === 0) {
    const pts = coords.map((c, i) => ({ east: c[0], north: c[1], name: provisional[i].name }));
    return { points: pts, residuals: [], sigma0: 0, iterations: 0 };
  }

  const sigDir = (sigmaDirSec / 3600) * (Math.PI / 180);
  const wDist = 1 / (sigmaDist * sigmaDist);
  const wDir = 1 / (sigDir * sigDir);

  const m = 2 * k;
  let iterations = 0;
  let lastV = new Array(m).fill(0);
  const W = new Array(m).fill(0);

  for (let iter = 0; iter < 5; iter++) {
    iterations += 1;
    const A: number[][] = Array.from({ length: m }, () => new Array(u).fill(0));
    const l = new Array(m).fill(0);

    for (let i = 0; i < k; i++) {
      const a = i;
      const b = i + 1;
      const [ea, na] = coords[a];
      const [eb, nb] = coords[b];
      const de = eb - ea;
      const dn = nb - na;
      const d = Math.hypot(de, dn);
      if (d < 1e-9) continue;
      const az = Math.atan2(de, dn);

      // distance observation row
      const r = 2 * i;
      if (col.has(b)) {
        A[r][col.get(b)! * 2] = de / d;
        A[r][col.get(b)! * 2 + 1] = dn / d;
      }
      if (col.has(a)) {
        A[r][col.get(a)! * 2] = -de / d;
        A[r][col.get(a)! * 2 + 1] = -dn / d;
      }
      l[r] = legs[i].distance - d;
      W[r] = wDist;

      // direction observation row
      const rb = 2 * i + 1;
      if (col.has(b)) {
        A[rb][col.get(b)! * 2] = dn / (d * d);
        A[rb][col.get(b)! * 2 + 1] = -de / (d * d);
      }
      if (col.has(a)) {
        A[rb][col.get(a)! * 2] = -dn / (d * d);
        A[rb][col.get(a)! * 2 + 1] = de / (d * d);
      }
      l[rb] = wrap((legs[i].bearing * Math.PI) / 180 - az);
      W[rb] = wDir;
    }

    // N = AᵀWA, rhs = AᵀWl
    const Nmat: number[][] = Array.from({ length: u }, () => new Array(u).fill(0));
    const rhs = new Array(u).fill(0);
    for (let p = 0; p < u; p++) {
      for (let q = 0; q < u; q++) {
        let sum = 0;
        for (let rIdx = 0; rIdx < m; rIdx++) sum += A[rIdx][p] * W[rIdx] * A[rIdx][q];
        Nmat[p][q] = sum;
      }
      let rsum = 0;
      for (let rIdx = 0; rIdx < m; rIdx++) rsum += A[rIdx][p] * W[rIdx] * l[rIdx];
      rhs[p] = rsum;
    }

    const dx = solveLinear(Nmat, rhs);

    // apply corrections
    for (const st of free) {
      const j = col.get(st)!;
      coords[st] = [coords[st][0] + dx[j * 2], coords[st][1] + dx[j * 2 + 1]];
    }

    // last_v = A·dx - l
    lastV = new Array(m).fill(0);
    for (let rIdx = 0; rIdx < m; rIdx++) {
      let s = 0;
      for (let p = 0; p < u; p++) s += A[rIdx][p] * dx[p];
      lastV[rIdx] = s - l[rIdx];
    }

    if (Math.max(...dx.map(Math.abs)) < 1e-6) break;
  }

  const dof = Math.max(m - u, 1);
  let vtwv = 0;
  for (let i = 0; i < m; i++) vtwv += lastV[i] * (W[i] * lastV[i]);
  const sigma0 = Math.sqrt(vtwv / dof);

  const points = coords.map((c, i) => ({ east: c[0], north: c[1], name: provisional[i].name }));
  const residuals: LsqResidual[] = legs.map((leg, i) => ({
    leg: i + 1,
    from: leg.from_name ?? null,
    to: leg.to_name ?? null,
    v_distance: lastV[2 * i],
    v_direction_sec: ((lastV[2 * i + 1] * 180) / Math.PI) * 3600,
    magnitude: Math.abs(lastV[2 * i]),
  }));

  return { points, residuals, sigma0, iterations };
}

File: services/engine/app/cogo/lsq.py
"""Least-squares traverse adjustment (parametric / observation-equation method).

Unknowns are the coordinates of the free stations. Each leg contributes two
observations — a distance and a direction — so a closed or link traverse has
redundancy 2 (the two closure conditions), which least squares distributes
optimally by minimising the weighted sum of squared residuals.

Open traverses have no redundancy, so least squares is not applicable there.
"""
from __future__ import annotations

import math
from dataclasses import dataclass

import numpy as np

from .geometry import Point


@dataclass
class LsqResult:
    points: list[Point]          # adjusted coordinates (station order, incl. fixed start)
    residuals: list[dict]        # per-leg distance & bearing residuals
    sigma0: float                # reference standard deviation (unitless a-posteriori)
    iterations: int


def _wrap(rad: float) -> float:
    while rad > math.pi:
        rad -= 2 * math.pi
    while rad < -math.pi:
        rad += 2 * math.pi
    return rad


def least_squares_adjust(
    start: Point,
    legs,                         # list with .bearing (deg), .distance (m)
    provisional: list[Point],    # provisional coords P0..Pk from raw traverse
    closed: bool,
    end_known: Point | None,
    sigma_dist: float = 0.01,
    sigma_dir_sec: float = 5.0,
) -> LsqResult:
    """Adjust a closed/link traverse by least squares.

    `provisional` are the unadjusted station coordinates (length = legs + 1).
    The first station is fixed; for a closed loop the last station is the start;
    for a link traverse the last station is `end_known`.
    """
    k = len(legs)
    coords = [np.array([p.east, p.north], dtype=float) for p in provisional]

    # Fixed stations: index 0 always; last index for closed/link.
    fixed = {0}
    if closed:
        coords[k] = np.array([start.east, start.north], dtype=float)
        fixed.add(k)
    elif end_known is not None:
        coords[k] = np.array([end_known.east, end_known.north], dtype=float)
        fixed.add(k)

    free = [i for i in range(k + 1) if i not in fixed]
    col = {st: j for j, st in enumerate(free)}  # station -> unknown block index
    u = 2 * len(free)
    if u == 0:
        # nothing to adjust
        pts = [Point(c[0], c[1], provisional[i].name) for i, c in enumerate(coords)]
        return LsqResult(points=pts, residuals=[], sigma0=0.0, iterations=0)

    sig_dir = math.radians(sigma_dir_sec / 3600.0)
    w_dist = 1.0 / (sigma_dist ** 2)
    w_dir = 1.0 / (sig_dir ** 2)

    iterations = 0
    last_v = np.zeros(2 * k)
    W = np.zeros(2 * k)
    for _ in range(5):
        iterations += 1
        A = np.zeros((2 * k, u))
        l = np.zeros(2 * k)
        for i, leg in enumerate(legs):
            a, b = i, i + 1
            ea, na = coords[a]
            eb, nb = coords[b]
            de, dn = eb - ea, nb - na
            d = math.hypot(de, dn)
            if d < 1e-9:
                continue
            az = math.atan2(de, dn)
            # distance observation row
            r = 2 * i
            if b in col:
                A[r, col[b] * 2] = de / d
                A[r, col[b] * 2 + 1] = dn / d
            if a in col:
                A[r, col[a] * 2] = -de / d
                A[r, col[a] * 2 + 1] = -dn / d
            l[r] = leg.distance - d
            W[r] = w_dist
            # direction observation row
            rb = 2 * i + 1
            if b in col:
                A[rb, col[b] * 2] = dn / (d * d)
                A[rb, col[b] * 2 + 1] = -de / (d * d)
            if a in col:
                A[rb, col[a] * 2] = -dn / (d * d)
                A[rb, col[a] * 2 + 1] = de / (d * d)
            l[rb] = _wrap(math.radians(leg.bearing) - az)
            W[rb] = w_dir

        Wm = np.diag(W)
        N = A.T @ Wm @ A
        rhs = A.T @ Wm @ l
        try:
            dx = np.linalg.solve(N, rhs)
        except np.linalg.LinAlgError:
            dx = np.linalg.lstsq(N, rhs, rcond=None)[0]
        # apply corrections
        for st in free:
            coords[st] = coords[st] + dx[col[st] * 2 : col[st] * 2 + 2]
        last_v = A @ dx - l
        if np.max(np.abs(dx)) < 1e-6:
            break

    m = 2 * k
    dof = max(m - u, 1)
    sigma0 = float(math.sqrt((last_v @ (W * last_v)) / dof))

    pts = [Point(float(c[0]), float(c[1]), provisional[i].name) for i, c in enumerate(coords)]
    residuals = []
    for i, leg in enumerate(legs):
        residuals.append({
            "leg": i + 1,
            "from": getattr(leg, "from_name", None),
            "to": getattr(leg, "to_name", None),
            "v_distance": float(last_v[2 * i]),
            "v_direction_sec": float(math.degrees(last_v[2 * i + 1]) * 3600.0),
            "magnitude": float(abs(last_v[2 * i])),
        })
    return LsqResult(points=pts, residuals=residuals, sigma0=sigma0, iterations=iterations)
